File: src/types/mame/detect_native_barcodes.ts
/**
 * TypeScript mirror of the ``mame.detect_native_barcodes`` RPC handler.
 *
 * Keep in sync with:
 *   - kuma_core/mame/ingest/demux.py  (NativeBarcodeUsage, detect_used_native_barcodes)
 *   - python-core/sidecar_mame/handlers/detect_native_barcodes.py (response dict)
 *   - python-core/sidecar_mame/models.py (DetectNativeBarcodesParams)
 */

/** Per-native-barcode read-volume usage entry. */
export interface NativeBarcodeUsage {
  /** MinKNOW barcode dir name (e.g. "barcode06"). */
  name: string
  /** Output subdir name derived from name (e.g. "sort_barcode06"). */
  sort_barcode_name: string
  /** Total on-disk size of all FASTQ(.gz) under the barcode dir, in bytes. */
  fastq_bytes: number
  /** fastq_bytes expressed in megabytes. */
  fastq_mb: number
  /** Fraction of total fastq_bytes across all native-barcode dirs. */
  share: number
  /** True when share meets min_share (this barcode was actually used). */
  is_used: boolean
}

/** Parameters for the mame.detect_native_barcodes RPC method. */
export interface DetectNativeBarcodesParams {
  /** Root of a MinKNOW run directory (must contain fastq_pass/). */
  minknow_run_dir: string
  /** Minimum share for a barcode to count as used. Default 0.05. */
  min_share?: number
}

/** Result of a mame.detect_native_barcodes RPC call. */
export interface DetectNativeBarcodesResult {
  /** Resolved path to the fastq_pass/ directory that was scanned. */
  fastq_pass: string
  /** Effective min_share threshold applied. */
  min_share: number
  /** Per-native-barcode usage entries, sorted by fastq_bytes descending. */
  native_barcodes: NativeBarcodeUsage[]
  /** Number of barcodes flagged is_used. */
  used_count: number
  /** Total number of native-barcode dirs detected. */
  total_count: number
}

File: python-core/sidecar_mame/handlers/detect_native_barcodes.py
"""``mame.detect_native_barcodes`` JSON-RPC handler.

Scans the ``fastq_pass/`` directory of a MinKNOW run and reports which native
barcodes were actually used, purely from on-disk FASTQ volume (a fast,
decompression-free proxy for read count).

RPC method name: ``mame.detect_native_barcodes``
Registered in ``sidecar_mame.dispatcher._METHODS`` (synchronous: stat-only,
fast, so it is NOT registered as an async method).

Parameter schema
----------------
See :class:`sidecar_mame.models.DetectNativeBarcodesParams` for full field
documentation and validation rules.

Response schema
---------------
``fastq_pass``      (str)   Resolved path to the scanned ``fastq_pass/`` dir.
``min_share``       (float) Threshold used for the ``is_used`` flag.
``native_barcodes`` (list)  Per-barcode summaries (name, sort_barcode_name,
                            fastq_bytes, fastq_mb, share, is_used).
``used_count``      (int)   Number of barcodes flagged as used.
``total_count``     (int)   Total native barcode dirs found.
"""

from __future__ import annotations

from pathlib import Path


def handle_detect_native_barcodes(params: dict) -> dict:
    """Detect which native barcodes were used in a MinKNOW run, from FASTQ volume.

    Parameters
    ----------
    params:
        Raw JSON-RPC params dict validated via
        :class:`~sidecar_mame.models.DetectNativeBarcodesParams`.

    Returns
    -------
    dict
        Result dict matching the response schema documented in the module
        docstring.
    """
    from sidecar_mame.models import DetectNativeBarcodesParams

    p = DetectNativeBarcodesParams.model_validate(params)

    from kuma_core.mame.ingest.demux import detect_used_native_barcodes
    from kuma_core.mame.ingest.sort_barcode import _nb_to_sort_barcode_name

    fastq_pass = Path(p.minknow_run_dir) / "fastq_pass"
    if not fastq_pass.is_dir():
        raise FileNotFoundError(
            f"fastq_pass/ directory not found under {p.minknow_run_dir}"
        )

    usages = detect_used_native_barcodes(fastq_pass, min_share=p.min_share)
    barcodes = [
        {
            "name": u.name,
            "sort_barcode_name": _nb_to_sort_barcode_name(u.name),
            "fastq_bytes": u.fastq_bytes,
            "fastq_mb": round(u.fastq_bytes / 1048576, 1),
            "share": round(u.share, 4),
            "is_used": u.is_used,
        }
        for u in usages
    ]
    return {
        "fastq_pass": str(fastq_pass.resolve()),
        "min_share": p.min_share,
        "native_barcodes": barcodes,
        "used_count": sum(1 for u in usages if u.is_used),
        "total_count": len(usages),
    }


__all__ = ["handle_detect_native_barcodes"]
